core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -621,8 +636,23 @@ private[log] class Cleaner(val id: Int,
       }
 
       // if we read bytes but didn't get even one complete message, our I/O buffer is too small, grow it and try again
-      if (readBuffer.limit() > 0 && result.messagesRead == 0)
-        growBuffers(maxLogMessageSize)
+      if (readBuffer.limit() > 0 && result.messagesRead == 0) {
+
+        // In some scenarios, a record could be bigger than the current maximum size configured for the log
+        //   1. A compacted topic using compression may contain a message set slightly larger than max.message.bytes
+        //   2. max.message.bytes of a topic could have been reduced after writing larger messages
+        // In these cases, grow the buffer to hold the next batch.
+        val maxSize = if (readBuffer.capacity >= maxLogMessageSize) {
+          checkCrc = true
+          val nextBatchSize = records.firstBatchSize
+          if (nextBatchSize == null)
+            throw new IllegalStateException(s"Could not determine next batch size for log ${sourceRecords.file} at position $position")
+          nextBatchSize.intValue
+        } else
+          maxLogMessageSize
```

Comment:
It seems that we need to grow buffers in a similar way in buildOffsetMapForSegment() too?